### ISSUE TITLE
Switch from PhantomJS to Chrome's headless mode

### ISSUE THREAD

**Issue by macbre** [macbre/phantomas#707]
PhantomJS 2.5 may **[never reach the stable state](https://news.ycombinator.com/item?id=13350728)** ([GitHub milestone](https://github.com/ariya/phantomjs/milestone/16) shows that it's 27% complete). Chrome has obviously more stable development team :wink: and leads the improve the web initiative - webp and spdy/http2 to name just a few.

Let's consider porting phantomas (and rename it to `chromas`? :smile: )  to use Chrome (and [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1338004)) run in a headless mode for performance auditing / monitoring.

> [Chrome vs PhantomJS vs SlimerJS](https://gist.github.com/macbre/03efa4479ab80dca4437#headlesschrome-via-puppeteer)

## Resources

* https://www.chromestatus.com/feature/5678767817097216 (since Chrome 59)
* https://developer.mozilla.org/en-US/Firefox/Headless_mode (since Firefox 55)
* https://chromedevtools.github.io/devtools-protocol/
* https://github.com/ChromeDevTools/debugger-protocol-viewer
* https://github.com/cyrus-and/chrome-remote-interface - sniffing HTTP traffic
* https://medium.com/@marco.luethy/running-headless-chrome-on-aws-lambda-fa82ad33a9eb
* https://www.aymen-loukil.com/en/blog-en/google-puppeteer-tutorial-with-examples/
* https://github.com/ariya/phantomjs/issues/15344 - PhantomJS repository has been archived
* https://docs.travis-ci.com/user/gui-and-headless-browsers/#using-the-chrome-addon-in-the-headless-mode

## Libraries

* ~https://github.com/adieuadieu/node-headless-chrome (work in progress, Windows support is missing)~
* https://github.com/GoogleChrome/puppeteer

## Puppeteer troubleshooting

* https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md#setting-up-chrome-linux-sandbox

## Initial work

Is being done in this gist - https://gist.github.com/macbre/fded35f86893bace033e3790079bee76

**Comment by OzzyCzech** [comment]
+1

**Comment by braincomb** [comment]
+1.  Chromium is twice as faster than PhantomJS.

**Comment by macbre** [comment]
[`v1.20.0`](https://github.com/macbre/phantomas/releases/tag/v1.20.0) was just tagged and pushed to npm. It's the last version of phantomas that's powered by PhantomJS.

Let's try what headless Chromium has to offer :)

**Comment by macbre** [comment]
Puppeteer's Chromium binary dependencies:

```
macbre@bjornoya:/tmp/node_modules/puppeteer/.local-chromium/linux-526987/chrome-linux$ ldd chrome | grep 'not found'
	libX11-xcb.so.1 => not found
	libXtst.so.6 => not found
	libXss.so.1 => not found
	libasound.so.2 => not found
	libatk-bridge-2.0.so.0 => not found
	libgtk-3.so.0 => not found
	libgdk-3.so.0 => not found
```

```
$ sudo apt-get install libx11-xcb1 libxtst6 libxss1 libasound2  libatk-bridge2.0-0  libgtk-3-0
```

**Comment by AdriwanKenoby** [comment]
I would like to help but I don't know where to start.

Add a selenium.js and a pupeteer.js to lib/engines/ but then I don't know what I have to refactor.

**Comment by AdriwanKenoby** [comment]
Could you provide more informations, about how to design the project, their is so much differences between phantomJS ans puppeteer.

`#!/usr/bin/env node

"use strict";

const puppeteer = require("puppeteer");

let argv = require("yargs")
	.usage("$0 <url>")
	.command(
		"$0 <url>",
		"collect metric for given url",
		(yargs) => {
			yargs.positional("url", {
				describe: "the url to use to collect metrics",
				type: "string"
			});
		},
		async (argv) => {
			const browser = await puppeteer.launch();
			const page = await browser.newPage();
			await page.goto(argv.url);
			let metrics = await page.metrics();
			console.log(metrics);
			await browser.close();
		}
	)
	.help().argv;
`

**Comment by AdriwanKenoby** [comment]
I'd like to help you with this project. Could you give me some advice to start with, I was thinking of using yargs to intuitively parse the command. The page object inherits directly from EventEmmiter and a lot of phantomas code is no longer needed. That's the big gap between phantomJS and puppeteer.

```
#!/usr/bin/env node

"use strict";

const puppeteer = require("puppeteer");

let argv = require("yargs")
	.usage("$0 <url>")
	.command(
		"$0 <url>",
		"collect metric for given url",
		(yargs) => {
			yargs.positional("url", {
				describe: "the url to use to collect metrics",
				type: "string"
			});
		},
		async (argv) => {
			const browser = await puppeteer.launch();
			const page = await browser.newPage();
			await page.goto(argv.url);
			let metrics = await page.metrics();
			console.log(metrics);
			await browser.close();
		}
	)
	.help().argv;
```

**Comment by macbre** [comment]
@AdriwanKenoby, thanks for your interest!

I'd definitely like to remove the extra code that was required to make PhantomJS work (and to be able to run SlimerJS as an engine as well) and be able to talk to Node.js layer (`ipc.js` sending JSON messages via `stderr` and passing events between both layers). Now we can focus on supporting just a single platform.

**Comment by macbre** [comment]
Guys,

It's been a while since the last activity in this thread. I want to give phantomas a new kick and refactor it to be more modular and use Puppeteer (headless Chromium) instead of PhantomJS (rest in peace, you've served us well!).

Here's my list of ideas, directions, notes:

* [x] Get rid of two rendering engines (SlimerJS and PhantomJS) and rely solely on headless Chromium (provided by Puppeteer package from Google)
* [ ] phantomas v2.0 will be a npm module only - no command line tool, various reporters, etc. Just a basic options handling, core modules and events emitting. Results will be returned as an object via Promise.
* [x] In [`phantomas-v2` branch](https://github.com/macbre/phantomas/compare/v2-headless?expand=1) I will port modules one by one and make sure tests pass
* [ ] Command line tool and various reporters will be moved to a new `macbre/phantomas-cli` repository
* [x] Provide `macbre/phantomas` docker container with CLI tool and reporters installed for an easy deployment

Some basic work has started here - https://gist.github.com/macbre/fded35f86893bace033e3790079bee76#file-script-js

@gmetais - your feedback on what will make your life with YellowLabTools easier are more than welcome here :)

**Comment by gmetais** [comment]
This is great news! (and it could lead to a v2.0 on YLT too!)
I'll help you as much as I can. Probably not on the core development, but by testing, adapting modules and later developing new modules.

One thing that would be great is switching offenders to a JSON format instead of a string. Which means that the CLI reporter will need to do the reverse job to display a string again. Or a string version of the offender could be kept in a JSON key.

I also consider removing from YLT the "Timeline" part, so YLT will stop spying the various jQuery functions. A JavaScript complexity index would still be cool, but a lighter one. Maybe Puppeteer gives access to the Chrome Timeline? Anyway, that's not a priority at all.

How can I help right now?

**Comment by macbre** [comment]
https://travis-ci.org/macbre/phantomas/jobs/473573512 - phantomas can now talk with Puppeteer's Chrome. And some tests do pass :)

Within a next week or two, all core modules should work as expected. Next step will be testing the npm module against various sites to find incompatibilities between PhantomJS and Chrome. @gmetais, this will be something that you can help with. And thanks!

```
✗ Errored » 138 honored ∙ 130 broken ∙ 18 errored (11.054s) 
```

**Comment by gmetais** [comment]
Surely I will.

**Comment by macbre** [comment]
Work in progress update:

* [x] `phantomas.incrMetric` and `phantomas.emit` called from page scope JS code are now passed to Node.js layer
* [x] test suite for `/dom-operations.html` now passes

**Comment by macbre** [comment]
Progress update:

* [x] more modules were ported to run in Chrome: `domMutations`, `localStorage`, `console`, `documentHeight`
* [x] `domComplexity` is almost refactored
* [x] - 43abdfa - offenders can now be checked in integration tests

**Comment by macbre** [comment]
Progress update:

* [x] phantomas now uses its custom user agent, i.e. `phantomas/2.0.0-beta (HeadlessChrome/72.0.3617.0)`
* [x] dc4231b - `jquery` module refactored, offenders are now better structured (key / value set instead of plain text)
* [x] a04766f - `events` module followed

**Comment by macbre** [comment]
Tests status so far:

✗ Broken » 254 honored ∙ 28 broken (13.168s)

**Comment by macbre** [comment]
More progress :)

✗ Broken » 272 honored ∙ 19 broken (42.727s)

**Comment by macbre** [comment]
```
✗ Broken » 279 honored ∙ 12 broken (12.621s) 
```

Now it's time to integrate `analyze-css` into phantomas v2.0.

**Comment by macbre** [comment]
`analyze-css` is back in 0712b40

```
✗ Broken » 315 honored ∙ 6 broken (15.264s)
```

**Comment by macbre** [comment]
118ce1b made all remaining tests pass :)

```
✓ OK » 323 honored (18.380s) 
```

**Comment by macbre** [comment]
966ae00 - TravisCI checks are green again :smiley:

**Comment by macbre** [comment]
```
✓ OK » 402 honored (35.484s) 
```

**Comment by macbre** [comment]
#726 has been merged into `phantomas-v2` branch. Now it's time to start working on `macbre/phantomas-cli`.

@gmetais, feel free to start testing using `examples/index.js` Node.js script. Thanks!

**Comment by macbre** [comment]
Documentation is now automatically generate from metadata JSON file that is built by parsing the code comments / annotations.

* [modules and metrics documentation](https://github.com/macbre/phantomas/blob/phantomas-v2/docs/metrics.md)
* [events](https://github.com/macbre/phantomas/blob/phantomas-v2/docs/events.md)

@gmetais - I hope that these resources will make developing on top of `phantomas` a bit easier task.

**Comment by gmetais** [comment]
Hi @macbre,
Sorry for the delay. Yes, thanks a lot for this new doc!

Should I add my comments about v2 to this post or create a new issue each time?

**Comment by gmetais** [comment]
I see you count brotli compressed files as gzipped (here: https://github.com/macbre/phantomas/blob/phantomas-v2/core/modules/requestsMonitor/requestsMonitor.js#L256)

For clarity, shouldn't Phantomas separate _gzipRequests_ and _brotliRequests_ and maybe add a new _compressedRequests_?

**Comment by gmetais** [comment]
Watch up here: https://github.com/macbre/phantomas/blob/phantomas-v2/core/modules/requestsMonitor/requestsMonitor.js#L303

When a request is answered as a redirect, entry.transferedSize is -1. So incrementing the _contentLength_ metric actually decreases it it by 1. Not a big deal, though!

What I find strange is that entry.responseSize is -1 too. The body size is zero, right, but headers are not.

**Comment by macbre** [comment]
#753 has been merged and [the first 2.0.0 npm pre-release package](https://www.npmjs.com/package/phantomas/v/2.0.0-alpha1) released.
